File: Striver/DP/Q46.py
# longest bitonic subsequence

def longestBitonicSubsequence(arr, n):
    dp1 = [1 for _ in range(n)]
    for ind in range(1, n):
        for prev_ind in range(ind):
            if arr[ind] > arr[prev_ind]:
                dp1[ind] = max(dp1[ind], 1 + dp1[prev_ind])

    dp2 = [1 for _ in range(n)]
    for ind in range(n-2, -1, -1):
        for prev_ind in range(ind+1, n):
            if arr[ind] > arr[prev_ind]:
                dp2[ind] = max(dp2[ind], 1 + dp2[prev_ind])

    mx = 1
    for i in range(n):
        mx = max(mx, dp1[i] + dp2[i] - 1)

    return mx

arr = [1, 11, 2, 10, 4, 5, 2, 1]
n = len(arr)
print(longestBitonicSubsequence(arr, n))

File: Striver/DP/Q46.java
package Striver.DP;

import java.util.Arrays;

public class Q46 {
    // longest bitonic subsequence

    public int longestBitonicSubsquence(int[] arr, int n){
        int[] dp1 = new int[n];
        Arrays.fill(dp1, 1);
        for(int i = 1; i < n; i++){
            for(int j = 0; j < i; j++){
                if(arr[i] > arr[j]){
                    dp1[i] = Math.max(dp1[i], dp1[j]+1);
                }
            }
        }

        int[] dp2 = new int[n];
        Arrays.fill(dp2, 1);
        for(int i = n-2; i >= 0; i--){
            for(int j = n-1; j > i; j--){
                if(arr[i] > arr[j]){
                    dp2[i] = Math.max(dp2[i], dp2[j]+1);
                }
            }
        }

        int max = 1;
        for(int i = 0; i < n; i++){
            max = Math.max(max, dp1[i] + dp2[i] - 1);
        }
        return max;
    }
}
